lib/kvdb/ikvdb.c

```diff
@@ -394,15 +424,40 @@ ikvdb_storage_add(const char *kvdb_home, struct kvdb_cparams *params)
         return err;
     }
 
+    err = ikvdb_is_pmem_only(kvdb_home, &meta, &pmem_only);
+    if (err)
+        return err;
+
+    if (!pmem_only && (params->storage.mclass[MP_MED_CAPACITY].path[0] != '\0')) {
+        log_err("Cannot add %s to a standard KVDB %s: capacity mclass path must be provided "
+                "at create time", params->storage.mclass[MP_MED_CAPACITY].path, kvdb_home);
+        return merr(EINVAL);
+    }
+
+    if (pmem_only && (params->storage.mclass[MP_MED_STAGING].path[0] != '\0') &&
+        (params->storage.mclass[MP_MED_CAPACITY].path[0] == '\0')) {
+        log_err("Cannot add %s to a pmem-only KVDB %s: capacity mclass must be added "
+                "before staging", params->storage.mclass[MP_MED_STAGING].path, kvdb_home);
+        return merr(EINVAL);
+    }
+
     for (i = MP_MED_BASE; i < MP_MED_COUNT; i++) {
-        if (i != MP_MED_CAPACITY && params->storage.mclass[i].path[0] != '\0') {
+        if (params->storage.mclass[i].path[0] != '\0') {
+            char buf[PATH_MAX];
             int j;
 
             if (meta.km_storage[i].path[0] != '\0') {
                 err = merr(EEXIST);
                 goto errout;
             }
 
+            err = kvdb_home_storage_path_get(kvdb_home, params->storage.mclass[i].path,
+                                             buf, sizeof(buf));
+            if (err)
+                goto errout;
+
+            strlcpy(params->storage.mclass[i].path, buf, sizeof(params->storage.mclass[i].path));
```

Comment:
Add a static assert that sizeof(buf) == sizeof(params->storage.mclass[MP_MED_BASE].path_